src/python/grpcio_tests/tests/testing/_time_test.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2017 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import time
+import random
+import unittest
+
+import grpc_testing
+
+_QUANTUM = 0.3
+_MANY = 10000
+
+
+class TimeTest(object):
+
+    def test_sleep_for(self):
+        start_time = self._time.time()
+        self._time.sleep_for(_QUANTUM)
+        end_time = self._time.time()
+
+        self.assertLessEqual(start_time + _QUANTUM, end_time)
+
+    def test_sleep_until(self):
+        start_time = self._time.time()
+        self._time.sleep_until(start_time + _QUANTUM)
+        end_time = self._time.time()
+
+        self.assertLessEqual(start_time + _QUANTUM, end_time)
+
+    def test_call_in(self):
+        event = threading.Event()
+
+        self._time.call_in(event.set, _QUANTUM)
+        self._time.sleep_for(_QUANTUM * 2)
+
+        self.assertTrue(event.is_set())
+
+    def test_call_at(self):
+        event = threading.Event()
+
+        self._time.call_at(event.set, self._time.time() + _QUANTUM)
+        self._time.sleep_for(_QUANTUM * 2)
+
+        self.assertTrue(event.is_set())
+
+    def test_cancel(self):
+        event = threading.Event()
+
+        future = self._time.call_in(event.set, _QUANTUM * 2)
+        self._time.sleep_for(_QUANTUM)
+        cancelled = future.cancel()
+        self._time.sleep_for(_QUANTUM * 2)
+
+        if cancelled:
+            self.assertFalse(event.is_set())
+            self.assertTrue(future.cancelled())
+        else:
+            self.assertTrue(event.is_set())
+
+    def test_many(self):
+        test_events = tuple(threading.Event() for _ in range(_MANY))
+        test_futures = {}
+        other_futures = []
+
+        for test_event in test_events:
+            test_futures[test_event] = self._time.call_in(
+                test_event.set, _QUANTUM * (2 + random.random()))
+        for _ in range(_MANY):
+            other_futures.append(
+                self._time.call_in(threading.Event().set, _QUANTUM * 1000 *
+                                   random.random()))
+        self._time.sleep_for(_QUANTUM)
+        cancelled = set()
+        for test_event, test_future in test_futures.items():
```

Comment:
I'm unfamiliar with Python3, and even less familiar with six's version gymnastics. I'd say leave it.